README
 INSTALLATION
===============

This file explains post compilation steps. It assumes that libnss-sqlite
is compiled and in system libraries directory (/lib).


 1. Create database
--------------------

libnss-sqlite use two SQLite databases to provide authentication services.
Thus the first thing you have to do is to create these DB. Depending on your
sources configuration, you may use one or two DB (--with-passwd-db &
--with-shadow-db switches). Default configuration uses 2 databases:
/etc/passwd.sqlite and /etc/shadow.sqlite, we'll use these here:

cd libnss-sqlite
sudo sqlite3 -init conf/passwd.sql /etc/passwd.sqlite
sudo sqlite3 -init conf/shadow.sql /etc/shadow.sqlite
sudo chmod o-r /etc/shadow.sqlite

That's all, databases are ready. Of course, it's up to you to populate them!
Each database contains a table named 'queries'. Each record inside this table
stores the query that should be performed in order to get the requested
information. Please, refer to conf/passwd.sql and conf/shadow.sql to get an
insight of the queries that can be customized and how to do it.

 2. Configure nsswitch.conf
----------------------------

The second step is to tell glibc that we use this lib. Open /etc/nsswitch.conf
and add sqlite at the end of the passwd, groups and shadow lines. You should
obtain something like :

# ...
passwd:         compat sqlite
group:          compat sqlite
shadow:         compat sqlite
# ...

 3. Test
---------

Insert some records into SQLite db tables, open a new shell (nsswitch.conf is
only read when a new application is launched) and try id [user in SQLite DB].

 4. Limitations
----------------

libnss-sqlite only handle users which are in its DB. You can't have an external
user linked to a DB stored group. This is because additional groups lookup use
username and is quite ugly to implement.


ChangeLog
2011-09-3 12:15 agamez
	* AUTHORS, README, groups.c, passwd.c, utils.c, shadow.c, utils.h
	  conf/passwd.sql, conf/shadow.sql: More deep refactorization of
	  previous code.
	* configure.ac: Set /etc as default directory for databases.

2011-08-30 17:22 agamez

	* AUTHORS, README, groups.c, passwd.c, utils.c, shadow.c, utils.h
          conf/passwd.sql, conf/shadow.sql: First stage of Alvaro Gamez
	  Machado's fork from original beuss' version. 
	  Query customization capability.
	  Non-essential bit of refactoring.

2007-08-17 17:32  beuss

	* AUTHORS, conf/shadow.sql, shadow.c: Integrating Sylvain Rochet's
	  patches to ensure uniqueness of username in
	  shadow table and proper structure initialization.

2007-07-10 19:11  beuss

	* INSTALL, README, conf/passwd.sql, conf/schema.sql,
	  conf/shadow.sql, config.h.in, configure.ac, groups.c, passwd.c,
	  shadow.c, utils.c, utils.h: Splitted DB in two files to allow
	  access control on shadow information.

2007-07-04 21:29  beuss

	* groups.c, passwd.c, utils.c: Added some missing finalize/close.
	  Removed some double or too early finalize/close.

2007-07-03 13:40  beuss

	* configure.ac: Updated version string.

2007-07-03 13:13  beuss

	* passwd.c, shadow.c: Finished first source documentation.

2007-07-03 12:58  beuss

	* INSTALL, README, configure.ac, nss-sqlite.h, utils.c: Added some
	  document (source, installation, configuration).
	  Removed CFLAGS stuff from configure.ac, it seems that it is not a
	  good thing to do.

2007-07-03 09:16  beuss

	* configure.ac: Modified configure.ac to remove -g from CFLAGS.

2007-07-03 08:58  beuss

	* Makefile.am: Modified Makefile.am to allow make dist to cleanly
	  build.

2007-07-03 08:49  beuss

	* bootstrap.sh, configure.ac: Added bootstrap script to prepare
	  build system.
	  Modified default prefix to install lib in /lib

2007-07-03 08:27  beuss

	* Makefile: Makefile is no more necessary.

2007-07-03 08:26  beuss

	* AUTHORS, COPYING, ChangeLog, INSTALL, Makefile, Makefile.am,
	  NEWS, README, config.h.in, configure.ac: Added autotools stuff.

2007-07-03 08:23  beuss

	* conf/schema.sql: Added indices on user_group. Speed up is awesone
	  (7min lookup with several thousands of groups vs 1 second).

2007-07-03 08:22  beuss

	* passwd.c: Removed unused include.

2007-07-02 19:02  beuss

	* .Makefile.in.swp, .groups.c.swp, .passwd.c.swp, .shadow.c.swp,
	  .utils.c.swp, .utils.h.swp, conf/.schema.sql.swp: Removed vim
	  swap files.

2007-07-02 18:58  beuss

	* .Makefile.in.swp, .groups.c.swp, .passwd.c.swp, .shadow.c.swp,
	  .utils.c.swp, .utils.h.swp, Makefile, conf, conf/.schema.sql.swp,
	  conf/schema.sql, groups.c, nss-sqlite.h, passwd.c, shadow.c,
	  utils.c, utils.h: Initial import.

2007-07-02 18:52  beuss

	* .: Initial layout.



utils.c
/*
 * Copyright (C) 2007, Sébastien Le Ray
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * utils.c : Some utility functions.
 */

#include "nss-sqlite.h"

#include <errno.h>
#include <grp.h>
#include <malloc.h>
#include <pwd.h>
#include <shadow.h>
#include <sqlite3.h>
#include <string.h>


/* Query the DB itself for the SQL query that is needed to resolve the call to getent function
 * @param pDb Database handle, will be closed if something fails.
 * @param getent_function The name of the getent function for which SQL statement is going to be retrieved.
 */
char *get_query(struct sqlite3* pDb, char *getent_function) {
    struct sqlite3_stmt* pSsql;
    const char* sql = "SELECT query FROM nss_queries WHERE name = ?";
    char *query;
    int res;

    if(sqlite3_prepare(pDb, sql, -1, &pSsql, NULL) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSsql);
        sqlite3_close(pDb);
        return NULL;
    }

    if(sqlite3_bind_text(pSsql, 1, getent_function, -1, SQLITE_STATIC) != SQLITE_OK) {
        NSS_DEBUG(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSsql);
        sqlite3_close(pDb);
        return NULL;
    }

    res = res2nss_status(sqlite3_step(pSsql), pDb, pSsql);
    if(res != NSS_STATUS_SUCCESS) {
        sqlite3_finalize(pSsql);
        sqlite3_close(pDb);
        return NULL;
    }

    query = strdup(sqlite3_column_text(pSsql, 0));
    sqlite3_finalize(pSsql);
    return query;
}

/*
 * Translate sqlite return code into a directly usable nss_status code.
 * @param pDb Database handle, will be closed if something fails.
 * @param pSt Statement to fetch from, will be finalized if something
 *      goes wrong.
 */

enum nss_status res2nss_status(int res, struct sqlite3* pDb, struct sqlite3_stmt* pSt) {
    switch(res) {
        /* Something was wrong with locks, try again later. */
        case SQLITE_BUSY:
            sqlite3_finalize(pSt);
            sqlite3_close(pDb);
            return NSS_STATUS_TRYAGAIN;
        /* No row returned (?) */

        case SQLITE_DONE:
            sqlite3_finalize(pSt);
            sqlite3_close(pDb);
            return NSS_STATUS_NOTFOUND;

        case SQLITE_ROW:
            return NSS_STATUS_SUCCESS;

        default:
            sqlite3_finalize(pSt);
            sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }
}

/*
 * Fill a group struct using given information.
 * @param pDb Handle to a database used to fetch group's members.
 * @param gbuf Struct which will be filled with various info.
 * @param buf Buffer which will contain all strings pointed to by
 *      gbuf.
 * @param buflen Buffer length.
 * @param name Groupname.
 * @param pw Group password.
 * @param gid Group ID.
 * @param errnop Pointer to errno, will be filled if something goes
 *      wrong.
 */

enum nss_status fill_group(struct sqlite3 *pDb, struct group *gbuf, char* buf, size_t buflen, struct group entry, int *errnop) {
    int name_length = strlen((char*)entry.gr_name) + 1;
    int pw_length = strlen((char*)entry.gr_passwd) + 1;
    int total_length = name_length + pw_length;
    int res;

    if(buflen < total_length) {
        *errnop = ERANGE;
        return NSS_STATUS_TRYAGAIN;
    }

    gbuf->gr_gid = entry.gr_gid;

    strcpy(buf, (const char*)entry.gr_name);
    gbuf->gr_name = buf;
    buf += name_length;

    strcpy(buf, (const char*)entry.gr_passwd);
    gbuf->gr_passwd = buf;
    buf += pw_length;

    /* We have a group, we now need to fetch its users */
    res = get_users(pDb, gbuf->gr_gid, buf, buflen - total_length, errnop);
    if(res == NSS_STATUS_SUCCESS) {
        gbuf->gr_mem = (char**)buf;
    }

    return res;
}

inline void fill_group_sql(struct group* entry, struct sqlite3_stmt* pSquery) {
    entry->gr_gid = sqlite3_column_int(pSquery, 0);
    entry->gr_name = sqlite3_column_text(pSquery, 1);
    entry->gr_passwd = sqlite3_column_text(pSquery, 2);

    return;
}



/*
 * Fill a passwd struct using given information.
 * @param pwbuf Struct which will be filled with various info.
 * @param buf Buffer which will contain all strings pointed to by
 *      pwbuf.
 * @param buflen Buffer length.
 * @param entry Passwd entry with needed data.
 * @param errnop Pointer to errno, will be filled if something goes wrong.
 */

enum nss_status fill_passwd(struct passwd* pwbuf, char* buf, size_t buflen, struct passwd entry, int* errnop) {
    int name_length = strlen(entry.pw_name) + 1;
    int pw_length = strlen(entry.pw_passwd) + 1;
    int gecos_length = strlen(entry.pw_gecos) + 1;
    int homedir_length = strlen(entry.pw_dir) + 1;
    int shell_length = strlen(entry.pw_shell) + 1;

    int total_length = name_length + pw_length + gecos_length + shell_length + homedir_length;

    if(buflen < total_length) {
        *errnop = ERANGE;
        return NSS_STATUS_TRYAGAIN;
    }

    pwbuf->pw_uid = entry.pw_uid;
    pwbuf->pw_gid = entry.pw_gid;

    strcpy(buf, entry.pw_name);
    pwbuf->pw_name = buf;
    buf += name_length;

    strcpy(buf, entry.pw_passwd);
    pwbuf->pw_passwd = buf;
    buf += pw_length;

    strcpy(buf, entry.pw_gecos);
    pwbuf->pw_gecos = buf;
    buf += gecos_length;

    strcpy(buf, entry.pw_dir);
    pwbuf->pw_dir = buf;
    buf += homedir_length;

    strcpy(buf, entry.pw_shell);
    pwbuf->pw_shell = buf;


    return NSS_STATUS_SUCCESS;
}

inline void fill_passwd_sql(struct passwd* entry, struct sqlite3_stmt* pSquery) {
    entry->pw_name = sqlite3_column_text(pSquery, 0);
    entry->pw_passwd = sqlite3_column_text(pSquery, 1);
    entry->pw_uid = sqlite3_column_int(pSquery, 2);
    entry->pw_gid =sqlite3_column_int(pSquery, 3);
    entry->pw_gecos = sqlite3_column_text(pSquery, 4);
    entry->pw_dir = sqlite3_column_text(pSquery, 5);
    entry->pw_shell = sqlite3_column_text(pSquery, 6);

    return;
}


/*
 * Fill an shadow password struct using given information.
 * @param spbuf Struct which will be filled with various info.
 * @param buf Buffer which will contain all strings pointed to by
 *      pwbuf.
 * @param buflen Buffer length.
 * @param name Username.
 * @param pw Encrypted password.
 * @param lstchg Date of last change (measured in days since 1970-01-01 00:00:00 +0000 (UTC))
 * @param min Min # of days between changes
 * @param max Max # of days between changes
 * @param warn # of days before password expires to warn user to change it
 * @param inact # of days after password expires until account is disabled
 * @param expire Date when account expires (measured in days since 1970-01-01 00:00:00 +0000 (UTC))
 * @param errnop Pointer to errno, will be filled if something goes wrong.
 */

enum nss_status fill_shadow(struct spwd *spbuf, char* buf, size_t buflen, struct spwd entry, int* errnop) {

    int name_length = strlen(entry.sp_namp) + 1;
    int pw_length = strlen(entry.sp_pwdp) + 1;

    if(buflen < name_length + pw_length) {
        *errnop = ERANGE;
        return NSS_STATUS_TRYAGAIN;
    }

    strcpy(buf, entry.sp_namp);
    spbuf->sp_namp = buf;
    buf += name_length;

    strcpy(buf, entry.sp_pwdp);
    spbuf->sp_pwdp = buf;

    spbuf->sp_lstchg = entry.sp_lstchg;
    spbuf->sp_min = entry.sp_min;
    spbuf->sp_max = entry.sp_max;
    spbuf->sp_warn = entry.sp_warn;
    spbuf->sp_inact = entry.sp_inact;
    spbuf->sp_expire = entry.sp_expire;

    return NSS_STATUS_SUCCESS;
}

inline void fill_shadow_sql(struct spwd* entry, struct sqlite3_stmt* pSquery) {
    entry->sp_namp = sqlite3_column_text(pSquery, 0);
    entry->sp_pwdp = sqlite3_column_text(pSquery, 1);
    entry->sp_lstchg = sqlite3_column_int(pSquery, 2);
    entry->sp_min = sqlite3_column_int(pSquery, 3);
    entry->sp_max = sqlite3_column_int(pSquery, 4);
    entry->sp_warn = sqlite3_column_int(pSquery, 5);
    entry->sp_inact = sqlite3_column_int(pSquery, 6);
    entry->sp_expire = sqlite3_column_int(pSquery, 6);

    return;
}


utils.h
/*
 * Copyright (C) 2007, Sébastien Le Ray
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef NSS_SQLITE_UTILS_H
#define NSS_SQLITE_UTILS_H

#include <sqlite3.h>
#include <grp.h>
#include <pwd.h>
#include <shadow.h>

char *get_query(struct sqlite3*, char*);

enum nss_status fill_passwd(struct passwd*, char*, size_t, struct passwd, int*);
inline void fill_passwd_sql(struct passwd*, struct sqlite3_stmt*);

enum nss_status fill_shadow(struct spwd*, char*, size_t, struct spwd, int*);
inline void fill_shadow_sql(struct spwd*, struct sqlite3_stmt*);

enum nss_status fill_group(struct sqlite3 *, struct group *, char*, size_t, struct group, int *);
inline void fill_group_sql(struct group*, struct sqlite3_stmt*);

#endif


groups.c
/*
 * Copyright (C) 2007, Sébastien Le Ray
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * groups.c : Functions handling groups entries retrieval.
 */
#include "nss-sqlite.h"
#include "utils.h"

#include <errno.h>
#include <grp.h>
#include <malloc.h>
#include <pthread.h>
#include <string.h>

/*
 * struct used to store data used by getgrent.
 */
static struct {
    sqlite3* pDb;
    sqlite3_stmt* pSt;
    int try_again;      /* flag to know if NSS_TRYAGAIN
                            was returned by previous call
                            to getgrent_r */
    /* group information cache used if NSS_TRYAGAIN was returned */
    struct group entry;
} grent_data = { NULL, NULL, 0, NULL};

/* mutex used to serialize xxgrent operation */
pthread_mutex_t grent_mutex = PTHREAD_RECURSIVE_MUTEX_INITIALIZER_NP;

/*
 * Free memory used by a dynamically allocated 2D table.
 * @param m Dynamically allocated table. Each element will
 * be freed, m will be freed too.
 * @param c Number of elements in t.
 */
static void free_2Dtable(char** t, int c) {
    int i;
    for(i = 0 ; i < c ; ++i) {
        free(t[i]);
    }
    free(t);
}

/*
 * Initialize grent functions (serial group access).
 */
enum nss_status _nss_sqlite_setgrent(void) {
    char* sql;
    pthread_mutex_lock(&grent_mutex);
    if(grent_data.pDb == NULL) {
        NSS_DEBUG("setgrent: opening DB connection\n");
        if(sqlite3_open(NSS_SQLITE_PASSWD_DB, &grent_data.pDb) != SQLITE_OK) {
            NSS_ERROR(sqlite3_errmsg(grent_data.pDb));
            return NSS_STATUS_UNAVAIL;
        }
        if(!(sql = get_query(grent_data.pDb, "setgrent")) ) {
            NSS_ERROR(sqlite3_errmsg(grent_data.pDb));
            sqlite3_close(grent_data.pDb);
            return NSS_STATUS_UNAVAIL;
        }
        if(sqlite3_prepare(grent_data.pDb, sql, -1, &grent_data.pSt, NULL) != SQLITE_OK) {
            NSS_ERROR(sqlite3_errmsg(grent_data.pDb));
            sqlite3_finalize(grent_data.pSt);
            sqlite3_close(grent_data.pDb);
            free(sql);
            return NSS_STATUS_UNAVAIL;
        }
    }
    free(sql);
    pthread_mutex_unlock(&grent_mutex);
    return NSS_STATUS_SUCCESS;
}

/*
 * Finalize grent functions.
 */
enum nss_status _nss_sqlite_endgrent(void) {
    NSS_DEBUG("endgrent: finalizing group serial access facilities\n");
    pthread_mutex_lock(&grent_mutex);
    if(grent_data.pDb != NULL) {
        sqlite3_finalize(grent_data.pSt);
        sqlite3_close(grent_data.pDb);
        grent_data.pDb = NULL;
    }
    pthread_mutex_unlock(&grent_mutex);
    return NSS_STATUS_SUCCESS;
}

/*
 * Return next group. see man getgrent_r
 * @param gbuf Buffer to store group data.
 * @param buf Buffer which will contain all string pointed
 * to by gbuf entries.
 * @param buflen buf length.
 * @param errnop Pointer to errno, will be filled if
 * an error occurs.
 */
enum nss_status
_nss_sqlite_getgrent_r(struct group *gbuf, char *buf,
                      size_t buflen, int *errnop) {
    int res;
    NSS_DEBUG("getgrent_r\n");
    pthread_mutex_lock(&grent_mutex);

    if(grent_data.pDb == NULL) {
        _nss_sqlite_setgrent();
    }

    if(grent_data.try_again) {
        res = fill_group(grent_data.pDb, gbuf, buf, buflen, grent_data.entry, errnop);
        /* buffer was long enough this time */
        if(res != NSS_STATUS_TRYAGAIN || (*errnop) != ERANGE) {
            grent_data.try_again = 0;
            pthread_mutex_unlock(&grent_mutex);
            return res;
        }
    }

    res = res2nss_status(sqlite3_step(grent_data.pSt), grent_data.pDb, grent_data.pSt);
    if(res != NSS_STATUS_SUCCESS) {
        grent_data.pDb = NULL;
        pthread_mutex_unlock(&grent_mutex);
        return res;
    }

    fill_group_sql(&grent_data.entry, grent_data.pSt);
    NSS_DEBUG("getgrent_r: fetched group #%d: %s\n", grent_data.entry.gr_gid, grent_data.entry.gr_name);

    res = fill_group(grent_data.pDb, gbuf, buf, buflen, grent_data.entry, errnop);
    if(res == NSS_STATUS_TRYAGAIN && (*errnop) == ERANGE) {
        /* cache result for next try */
        grent_data.try_again = 1;

        pthread_mutex_unlock(&grent_mutex);
        return NSS_STATUS_TRYAGAIN;
    }
    pthread_mutex_unlock(&grent_mutex);
    return NSS_STATUS_SUCCESS;
}

/**
 * Get group by name.
 * @param name Groupname.
 * @param buf Buffer which will contain all string pointed
 * to by gbuf entries.
 * @param buflen buf length.
 * @param errnop Pointer to errno, will be filled if
 * an error occurs.
 */

enum nss_status
_nss_sqlite_getgrnam_r(const char* name, struct group *gbuf,
                      char *buf, size_t buflen, int *errnop) {
    sqlite3 *pDb;
    struct sqlite3_stmt* pSt;
    struct group entry;
    int res;
    char* sql;

    NSS_DEBUG("getgrnam_r : looking for group %s\n", name);

    if(sqlite3_open(NSS_SQLITE_PASSWD_DB, &pDb) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(!(sql = get_query(pDb, "getgrnam_r")) ) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_prepare(pDb, sql, -1, &pSt, NULL) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_bind_text(pSt, 1, name, -1, SQLITE_STATIC) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    res = res2nss_status(sqlite3_step(pSt), pDb, pSt);
    if(res != NSS_STATUS_SUCCESS) {
        free(sql);
        return res;
    }

    fill_group_sql(&entry, pSt);

    res = fill_group(pDb, gbuf, buf, buflen, entry, errnop);

    sqlite3_finalize(pSt);
    sqlite3_close(pDb);
    free(sql);
    return res;
}

/*
 * Get group by GID.
 * @param gid GID.
 * @param buf Buffer which will contain all string pointed
 * to by gbuf entries.
 * @param buflen buf length.
 * @param errnop Pointer to errno, will be filled if
 * an error occurs.
 */

enum nss_status
_nss_sqlite_getgrgid_r(gid_t gid, struct group *gbuf,
                      char *buf, size_t buflen, int *errnop) {
     sqlite3 *pDb;
     struct sqlite3_stmt* pSt;
     struct group entry;
     int res;
     char* sql;


    NSS_DEBUG("getgrgid_r : looking for group #%d\n", gid);

    if(sqlite3_open(NSS_SQLITE_PASSWD_DB, &pDb) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(!(sql = get_query(pDb, "getgrgid_r")) ) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_prepare(pDb, sql, -1, &pSt, NULL) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_bind_int(pSt, 1, gid) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    res = res2nss_status(sqlite3_step(pSt), pDb, pSt);
    if(res != NSS_STATUS_SUCCESS) {
        free(sql);
        return res;
    }

    fill_group_sql(&entry, pSt);

    res = fill_group(pDb, gbuf, buf, buflen, entry, errnop);

    sqlite3_finalize(pSt);
    sqlite3_close(pDb);
    free(sql);
    return res;

}

/*
 * Haven't seen any detailled documentation about this function.
 * Anyway it have to fill in groups for the specified user without
 * adding his main group (group param).
 * @param user Username whose groups are wanted.
 * @param group Main group of user (should not be put in groupsp).
 * @param start Index from which groups filling must begin (initgroups_dyn
 * is called for every backend). Can be updated
 * @param size Size of groups vector. Can be modified if function needs
 * more space (should not exceed limit).
 * @param groupsp Pointer to the group vector. Can be realloc'ed if more
 * space is needed.
 * @param limit Max size of groupsp (<= 0 if no limit).
 * @param errnop Pointer to errno (filled if an error occurs).
 */

enum nss_status
_nss_sqlite_initgroups_dyn(const char *user, gid_t gid, long int *start,
                          long int *size, gid_t **groupsp, long int limit,
                                                    int *errnop) {
    sqlite3 *pDb;
    struct sqlite3_stmt *pSt;
    char* sql;
    int res;
    NSS_DEBUG("initgroups_dyn: filling groups for user : %s, main gid : %d\n", user, gid);

    if(sqlite3_open(NSS_SQLITE_PASSWD_DB, &pDb) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(!(sql = get_query(pDb, "initgroups_dyn")) ) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_prepare(pDb, sql, -1, &pSt, NULL) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_bind_text(pSt, 1, user, -1, SQLITE_STATIC) != SQLITE_OK) {
        NSS_ERROR("Unable to bind username in initgroups_dyn\n");
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_bind_int(pSt, 2, gid) != SQLITE_OK) {
        NSS_ERROR("Unable to bind gid in initgroups_dyn\n");
        sqlite3_finalize(pSt);
        sqlite3_close(pDb);
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    res = res2nss_status(sqlite3_step(pSt), pDb, pSt);
    if(res != NSS_STATUS_SUCCESS) {
        free(sql);
        return res;
    }

    do {
        int gid = sqlite3_column_int(pSt, 0);
        NSS_DEBUG("initgroups_dyn: adding group %d\n", gid);
        /* Too short, doubling size */
        if(*start == *size) {
            if(limit > 0) {
                if(*size < limit) {
                    *size = (limit < (*size * 2)) ? limit : (*size * 2);
                } else {
                    /* limit reached, tell caller to try with a bigger one */
                    NSS_ERROR("initgroups_dyn: limit was too low\n");
                    *errnop = ERANGE;
                    sqlite3_finalize(pSt);
                    sqlite3_close(pDb);
                    free(sql);
                    return NSS_STATUS_TRYAGAIN;
                }
            } else {
                (*size) = (*size) * 2;
            }
            *groupsp = realloc(*groupsp, sizeof(**groupsp) * (*size));
        }
        (*groupsp)[*start] = gid;
        (*start)++;
        res = sqlite3_step(pSt);
    } while(res == SQLITE_ROW);
    *groupsp = realloc(*groupsp, sizeof(**groupsp) * (*start));
    *size = *start;

    sqlite3_finalize(pSt);
    sqlite3_close(pDb);
    free(sql);

    return NSS_STATUS_SUCCESS;
}

/*
 * Fills all users for a given group.
 * @param buffer Buffer which will contain all users' names headed
 * with a char* pointers area containing pointer to members' names,
 * ending by NULL.
 * @param pDb DB handle to fetch users (must be opened).
 * @param gid GID.
 * @param buflen Buffer length.
 * @param errnop Pointer to errno, will be filled if an error occurs.
 */

enum nss_status get_users(sqlite3* pDb, gid_t gid, char* buffer, size_t buflen, int* errnop) {
    struct sqlite3_stmt *pSt;
    int res, msize = 20, mcount = 0, i, ptr_area_size;
    char* sql;
    char* next_member;
    char **members;
    char **ptr_area = (char**)buffer;

    NSS_DEBUG("get_users: looking for members of group #%d\n", gid);
    
    if(!(sql = get_query(pDb, "get_users")) ) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_close(pDb);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_prepare(pDb, sql, strlen(sql), &pSt, NULL) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        return NSS_STATUS_UNAVAIL;
    }

    if(sqlite3_bind_int(pSt, 1, gid) != SQLITE_OK) {
        NSS_ERROR(sqlite3_errmsg(pDb));
        sqlite3_finalize(pSt);
        return NSS_STATUS_UNAVAIL;
        free(sql);
    }

    res = sqlite3_step(pSt);

    if(res != SQLITE_ROW) {
        sqlite3_finalize(pSt);
        if(res == SQLITE_DONE) {
            NSS_DEBUG("get_users: No member found\n");
            if(buflen < sizeof(char*)) {
                *errnop = ERANGE;
                free(sql);
                return NSS_STATUS_TRYAGAIN;
            }
            ptr_area[0] = NULL;
            free(sql);
            return NSS_STATUS_SUCCESS;
        }
        free(sql);
        return NSS_STATUS_UNAVAIL;
    }

    /* members is a buffer to temporary hold members (we need to know the count
     * before going further) and SQLite doesn't seems to offer such a
     * functionnality. */
    members = (char**)malloc(msize * sizeof(char*));
    do {
        const unsigned char* member = sqlite3_column_text(pSt, 0);

        if(msize == mcount) {
            msize *= 2;
            members = (char**)realloc(members, msize * sizeof(char*));
        }
        members[mcount] = strdup((char*)member);
        ++mcount;
        res = sqlite3_step(pSt);
    } while(res == SQLITE_ROW);

    free(sql);
    sqlite3_finalize(pSt);

    /* Here is what we want to get :
     * __________________________________________________
     * ...|@1|@2|@3|...|NULL|member1|member2|member3|...
     * --------------------------------------------------
     *    ^ gr_mem
     */

    /* Let's build addresses part */
    ptr_area_size = (mcount + 1) * sizeof(char *);

    if(buflen < ptr_area_size) {
        free_2Dtable(members, mcount);
        (*errnop) = ERANGE;
        return NSS_STATUS_TRYAGAIN;
    }

    next_member = buffer + ptr_area_size;
    buflen -= ptr_area_size;
    for(i = 0 ; i < mcount ; ++i) {
        int l = strlen(members[i]) + 1;
        if(buflen < l) {
            free_2Dtable(members, mcount);
            (*errnop) = ERANGE;
            return NSS_STATUS_TRYAGAIN;
        }
        strcpy(next_member, members[i]);
        ptr_area[i] = next_member;
        buflen -= l;
        next_member  += l;
    }
    ptr_area[i] = NULL;
    free_2Dtable(members, mcount);
    return TRUE;
}

